File: desktop/src/lib/api/anime.ts
import { invoke } from '@tauri-apps/api/core'
import type { AnimeSummary, AnimeDetails, DiscoverFilters, DiscoverPage } from '$lib/types/anime'

export const HOME_GENRES = ['Action', 'Adventure', 'Romance', 'Fantasy', 'Comedy']

export async function getTrending(): Promise<AnimeSummary[]> {
    return invoke('get_trending')
}

export async function getPopular(perPage = 20): Promise<AnimeSummary[]> {
    return invoke('get_popular', { perPage })
}

export async function getTopRated(): Promise<AnimeSummary[]> {
    return invoke('get_top_rated')
}

export async function getHomeGenres(): Promise<Record<string, AnimeSummary[]>> {
    return invoke('get_home_genres', { genres: HOME_GENRES })
}

export async function getAnimeDetails(id: number): Promise<AnimeDetails> {
    return invoke('get_anime_details', { id })
}

export async function searchAnime(query: string, perPage = 30): Promise<AnimeSummary[]> {
    return invoke('search_anime', { query, perPage })
}

export async function discoverAnime(
    filters: DiscoverFilters = {},
    page = 1,
    perPage = 24,
): Promise<DiscoverPage> {
    return invoke('discover_anime', { filters, page, perPage })
}

File: desktop/src-tauri/src/commands/anime.rs
use crate::state::AppState;
use namizo_core::{AnimeDetails, AnimeSummary, DiscoverFilters, DiscoverPage};
use std::collections::HashMap;
use tauri::State;

#[tauri::command]
pub async fn get_trending(state: State<'_, AppState>) -> Result<Vec<AnimeSummary>, String> {
    state.anime_service.trending(20).await
}

#[tauri::command]
pub async fn get_popular(
    per_page: Option<u8>,
    state: State<'_, AppState>,
) -> Result<Vec<AnimeSummary>, String> {
    state.anime_service.popular(per_page.unwrap_or(20)).await
}

#[tauri::command]
pub async fn get_top_rated(state: State<'_, AppState>) -> Result<Vec<AnimeSummary>, String> {
    state.anime_service.top_rated(20).await
}

#[tauri::command]
pub async fn get_home_genres(
    genres: Vec<String>,
    state: State<'_, AppState>,
) -> Result<HashMap<String, Vec<AnimeSummary>>, String> {
    state.anime_service.genres_batch(genres).await
}

#[tauri::command]
pub async fn get_anime_details(
    id: u32,
    state: State<'_, AppState>,
) -> Result<AnimeDetails, String> {
    state.anime_service.details(id).await
}

#[tauri::command]
pub async fn search_anime(
    query: String,
    per_page: Option<u8>,
    state: State<'_, AppState>,
) -> Result<Vec<AnimeSummary>, String> {
    state.anime_service.search(&query, per_page.unwrap_or(30)).await
}

#[tauri::command]
pub async fn discover_anime(
    filters: Option<DiscoverFilters>,
    page: Option<u32>,
    per_page: Option<u8>,
    state: State<'_, AppState>,
) -> Result<DiscoverPage, String> {
    state
        .anime_service
        .discover(filters.unwrap_or_default(), page.unwrap_or(1), per_page.unwrap_or(24))
        .await
}